onnxruntime/core/session/inference_session.cc

```diff
@@ -1733,59 +1732,87 @@ common::Status InferenceSession::Run(IOBinding& io_binding) {
+  onnxruntime::contrib::OrtTasks::GetInstance().CreateBackgroundTask(run_id);
+
+  LOGS(*session_logger_, WARNING) << "Session::Forward" << run_id;
 
-  // Wait for events from
+  // background task is setup, unblock background thread to continue
+  setup_promise.set_value();
+
+  LOGS(*session_logger_, WARNING) << "after setup_promise.set_value()" << run_id;
+
+  // Wait for data/signal from
   // 1. Yield op, if the bg thread sucessfully reached Yield's signal point
   // 2. The end of bg thread, if it hit execptions and returned earlier
-  onnxruntime::contrib::OrtEventPool::GetInstance().WaitAndResetEvent(main_thread_event_id);
+  auto forward_outputs = onnxruntime::contrib::OrtTasks::GetInstance().WaitForForwardOutputs(run_id);
+  Status forward_status = forward_outputs.first;
+  user_outputs = forward_outputs.second;
+
+  LOGS(*session_logger_, WARNING) << "after WaitForForwardOutputs" << run_id;
 
   // background thread has completed without hitting Yield Op
-  if (task_.bg_thread_future_.wait_for(std::chrono::milliseconds(0)) == std::future_status::ready) {
-    Status bg_thread_status = task_.bg_thread_future_.get();
-    task_.bg_thread_.join();
-    return bg_thread_status;
+  if (!forward_status.IsOK()) {
+    bg_threads_[run_id].join();
+    return forward_status;
   }
 
-  onnxruntime::contrib::OrtMessageQueue::GetInstance().PopAll(user_outputs);
   return Status::OK();
 }
 
-common::Status InferenceSession::ContinueRunInBackground(const std::vector<OrtValue>& backward_output_grads) {
-  for (const auto& ort_value : backward_output_grads) {
-    onnxruntime::contrib::OrtMessageQueue::GetInstance().Push(ort_value);
-  }
+common::Status InferenceSession::ContinueRunInBackground(int64_t run_id, const std::vector<OrtValue>& backward_output_grads) {
+  LOGS(*session_logger_, WARNING) << "Session::Backward" << run_id;
 
   // resume background thread
-  const int64_t background_thread_event_id = 1;
-  onnxruntime::contrib::OrtEventPool::GetInstance().SignalEvent(background_thread_event_id);
+  onnxruntime::contrib::OrtTasks::GetInstance().SetBackwardInputs(run_id, backward_output_grads, false);
+
+  Status bg_thread_status = onnxruntime::contrib::OrtTasks::GetInstance().WaitForStatus(run_id);
 
-  Status bg_thread_status = task_.bg_thread_future_.get();
   // wait for bg_thread to complete
-  if (task_.bg_thread_.joinable()) {
-    task_.bg_thread_.join();
+  if (bg_threads_[run_id].joinable()) {
+    bg_threads_[run_id].join();
   }
 
   return bg_thread_status;
 }
+
+void InferenceSession::CancelBackgroundTask(int64_t run_id) {
+  LOGS(*session_logger_, WARNING) << "CancelBackgroundTask " << run_id;
+
+  // resume background thread with terminate = true
+  onnxruntime::contrib::OrtTasks::GetInstance().SetBackwardInputs(run_id, {}, true);
+
+  // wait for bg_thread to complete
+  if (bg_threads_[run_id].joinable()) {
+    bg_threads_[run_id].join();
+  }
+
+  onnxruntime::contrib::OrtTasks::GetInstance().RemoveTask(run_id);
```

Comment:
If the task has been removed, should it also be removed from ```bg_threads_```? Or does a reference keep living in that map?
If we remove the entry from the map, we would also need a protect the map with a locking mechanism?

Reply:
done.